torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -225,8 +225,16 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
     // `detach_` from `zero_grad`, which is incompatible with views.
     TORCH_INTERNAL_ASSERT(!grad.is_alias_of(bucket_view));
     TORCH_INTERNAL_ASSERT(grad.device() == bucket_view.device());
-    TORCH_INTERNAL_ASSERT(grad.numel() == bucket_view.numel());
-    bucket_view.copy_(grad.view({-1}), /* non_blocking */ true);
+    TORCH_INTERNAL_ASSERT(grad.sizes() == bucket_view.sizes());
+    // Checks below are too harsh.  AccumulateGrad doesn't HAVE to obey
+    // the grad layout contract.  The penalty for disobedience is reduced
+    // performance, not numerical death.
+    // What should I do here?  Warn?  Do nothing?
+    TORCH_INTERNAL_ASSERT(grad.strides() == bucket_view.strides());
+    TORCH_INTERNAL_ASSERT(variable.is_non_overlapping_and_dense() ?
+                         (grad.strides() == variable.strides()) :
+                         grad.is_contiguous(at::MemoryFormat::Contiguous));
+    bucket_view.copy_(grad, /* non_blocking */ true);
```

Comment:
I think you lost the `view(-1)` here.

Reply:
Nope, deliberately got rid of it, that's the point :)  grad.view(-1) will fail if grad has permuted strides.  Here bucket_view has the same sizes as grad and _probably_ the same layout, so the copy will always be functionally correct and _probably_ performant (coalesced reads and writes) regardless of how grad's strides are permuted.

Reply:
Ho right these new buckets are introduced below. Looks good !